# Cryptographie : chiffrement affine

## 1. Correspondance numérique / alphabétique

|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|0|1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|

In [2]:
alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

### 1.1 Exercice : passage à l'alphabétique

In [5]:
def alpha(m: list) -> str:
    decript = ""
    for i in range(len(m)):
        for j in range(len(alphabet)):
            if m[i] == j:
                decript += f"{alphabet[j]}"
    return decript

In [6]:
alpha([18,0,11,20,19])

'SALUT'

### 1.2 Exercice : passage au numérique

In [7]:
def num(m: int) -> list:
    decript = []
    for i in range(len(m)):
        for j in range(len(alphabet)):
            if m[i] == alphabet[j]:
                decript += [j]
    return decript

In [8]:
num("SALUT")

[18, 0, 11, 20, 19]

## 2. Chiffrement affine

- Fonction de chiffrement : **ek(m) = a x m + b**
- Fonction de déchiffrement : **dk(c) = (a)-1 x (c - b)**

### 2.1 Exercice : chiffrement

In [21]:
def chiffre(m: str, a: int, b: int) -> str:
    liste = num(m)
    for i in range(len(liste)):
        liste[i] = (a * liste[i]) + b
        while liste[i] >= 26:
            liste[i] -= 26
    return alpha(liste)

In [22]:
chiffre("IUT", 3, 2)

'AKH'

### 2.2 Exercice : inverse

In [76]:
def inverse(x: int, n: int):
    a = -1
    while a != n:
        a += 1
        if (x * a) % n == 1:
            b = a
            break
        elif a == n - 1:
            b = "pas inversible"
    return b

In [77]:
inverse(3,26)

9

In [78]:
inverse(4,26)

'pas inversible'

### 2.3 Exercice : déchiffremment

In [90]:
def dechiffre(c: str, a: int, b: int) -> str:
    liste = num(c)
    for i in range(len(liste)):
        liste[i] = inverse(a,26) * (liste[i] - b)
        while liste[i] >= 26:
            liste[i] -= 26
        while liste[i] < 0:
            liste[i] += 26
    return alpha(liste)

In [91]:
dechiffre("AKH",3,2)

'IUT'

In [92]:
c="EYCPRTYDWYNWTWQPRJWIQPRUIOTYCZZYQWOOYJFWHUYTUPRWKCJJWJ"
dechiffre(c,3,2)

'SQANFXQJYQVYXYWNFLYCWNFGCEXQAZZQWYEEQLBYTGQXGNFYUALLYL'

## 3. Bonus : analyse fréquentielle

|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|9.42|1.02|2.64|3.39|15.87|0.95|1.04|0.77|8.41|0.89|0.0|5.34|3.24|7.15|5.14|2.86|1.06|6.46|7.90|7.26|6.24|2.15|0.0|0.3|0.24|0.32|

### 3.1 calcul de la fréquence d'apparition

In [147]:
def frequence(c: str) -> list:
    liste = 26*[0]
    for i in range(len(c)):
        for j in range(len(alphabet)):
            if c[i] == alphabet[j]:
                liste[j] += 1
    for k in range(len(liste)):
        liste[k] = round(liste[k]/len(c)*100,2)
    return liste

In [148]:
message="OYQVARYYBWGAYTYIAYLLYOSFSHYEJWLTTYFESQQYEAPQBEVSEAPWBYLERVEBELVSLTARYJEQYOEONYYHYQLYJYHHYQWRVYMBWLVYHYALBLYMSYLYFSOVWSLYQALHYMEHYDSGAYYMBSLYIEHEOVSGAYEAOWALQTYHEJEVESHHYTYQYQBSWRQLYJYHHYQWRVLYAQQSETYLWJYLHYQBHERQQYOLYVQTYHELMYEJQWHAYTYHYMBSLYHYVWSHYTYHEMWLVARYQVEVSWRQBEVSEHYJHSRTYYTWVYYTARYGASBYMYRVEQQYZBASQQERVBWALERRSNSHYLARYBHERYVYVWAVYRVSYLYBWALQASFSYBELTYQQJSLYQQSRSQVLYQTYHYMBSLYHEBLSROYQQYHYSEDSHYFYLQQEJEQYTERQQWRFESQQYEAOWQMSGAYBWLVYAQYTYQBHERQFWHYQEHYRRYMSGASBWALLWRVQEAFYLQWRBYABHYYVLYQVEALYLHEHSJYLVYTERQHEIEHEPSY"
print(frequence(message))

[5.88, 4.55, 0.0, 0.38, 8.73, 1.52, 1.14, 7.02, 0.57, 2.09, 0.0, 7.78, 2.09, 0.38, 1.9, 0.57, 9.68, 5.31, 7.4, 3.42, 0.0, 5.5, 4.74, 0.0, 19.17, 0.19]


### 3.2 déchiffrement du message codé

- E a été chiffré en 8.73
- A a été chiffré en 5.88
- Il y a donc + de E que de A, la clé change en fonction de la fréquence de la lettre